Cesium3DStyle Case Error

srothst1 opened the pull request:
Fixes #9418

Updated `Cesium3DTileBatchTable.prototype.setColor` in Cesium3DTileBatchTable.js to return if a color is not defined, rather than setting an undefined color.

Modified Cesium3DTileBatchTableSpec.js to correctly test the update to `Cesium3DTileBatchTable.prototype.setColor`.

Updated the `Cesium3DTileset` documentation explaining the new default color value if no color conditions are given.

[Here ](http://localhost:8080/Apps/Sandcastle/index.html#c=xVj/b9s2Fv9XHnyHWca5Uhyn6+rZwTVJ71pcuhRzsP1QBwdaom0iNOmRlDPd5v99j5SoL7bqeGiHCwLYovg+78P3nY4ieCefwEhINQWzojC8gXvGqYapyTgTS+BELFOypHaTxjUKTCRsy5KUcFhQYlJFdR84e6QwTxlPUEiHMxFFhxB6Q2O2YDExTIoRrIzZ6FEULZlZpfMwluvommqWrv89jYbJC2N5REZRGq2JNlRFDfmogJ+JLVGwZfSJKpiAoE+Qo4Q/ubVg1ond87UUhjBB1azTh99mAgAxFa58VBLPQ9XIC8YKz0V/loon9/mWoNefiV3v+1zZioiE72ubohQV0w2J6dstFeZdvinImYU6poKGMUEAbXGchd4kSSEOd9MPlf2ssa03nBAQDTJVQH8l6w2vXERNmNORen3lJYs3SK1xlrvalsAdo85qo9iaGbalOiRJErTglYyniG2ZMYECGAExWVNFLF8u5SMQgzuIMZzuqcj3hQhlvRI48ydUGyaKYCjoXhNl8BsRw3Ch5PqGLtGqOngxOD8Phxevzvtw8Sp8+fq7PgxfnVmnAEjF0NwFjAMGWFFi6ZewH4hZhUb+iKtE6GBQiAJsmIlXn9v2wu/bee834nqRithq1cX6teRSwTyDNVpcOeusaPyoYYHLIuUctoSnmFuaiZiCkAYI50608vyKbPNMLEE2Sm6oMhl622uE2Kq6yj4Ue4JefvAWz4V50jZCNf8Y3tgt9jA0KMyW0AUmiC7NCCWLEcw6f/+tyPdPXc935N93H3aYVrnUrvh0JGtYscTK4Qw2gk9+EfArInucHUwmE2csm6WYuhYj6D6tmKHd3qzz0D8q2F1yonW3IasfszmavduHs/DlCRBIE5MG1TVQMA6zUwjMFYsfm6K2YBKhaHKKvDZS7KnmbLkymuO+U0msqUGXHIKcKo/VlrbJu3VnzD2QKEIYo1L6Oa/hDrhfMQ34b6Obcmw4MdHUYzw0gscl287mVRnyKyTgSLzh/EcM4ASTHhnXytrXyYGTo7b8Cu25gSmRW5NssKitka7uwu+/w7GtqjpYt8wnr6OwapwR4YL5da+xpeFU64vj+fOswfVKPl1l3pb32YYG89pDYW79ZCsotL0C52AkIRfYtDE0Rl73l/rI/ll6o2csnyuuW9KdM/86x874+H2DaOWp/wPZWpg8TxgrNUm5GR2+2xV+LBuSTbeqv8zxpAmge3GZKUiYNsR2I9ttgYCmnMYGd2DXNgq7D5f5uHXYfG4K0Xt5LaVKbCenwQarH01uUcSkCUZg8SzF0i38ZX3Kn2NU2XzW8YvBlsbnQc32+UD4N+5poQ/60PK+OAa+7vXBgcw68I9Kxd7p6q+sC9v2FpDNrb3eFzVPf84dXGJZOBtcNKpwW71uFTvbb3hUnCJ39m1DLqOcy6eTBM8GDcn2DnmsqzxXyzAHfl5RAQmOUIl8EiA3LoKxCdnpa9bJM+QqAx/McC9xgi1y4NbHvnUOYim6lsV0xunCQMzRpzgKb1ID1M78btLLU8gOiORYFuVgVep8RCwUuhO3CH1tkX1HS2Sc2roQ/pJSleXspL3XhEws5EciKA5LvTyBwi3TbM44Mxnm0qyzYgn2k1nHlYbi3hLmqt9b3m8cmeAzdxhbzMPbt/+6/+/17fvr/9Ttaq8u+Zk5mg5HfOVuLaWlkXBamBvn4690BHvbccCTyihLat5yar9eZe8ThPQUENBN7FYgLJh8OnsI8cPpR5DSGd7UpzjFmfJgALemOVQ3OK4O7XiTF3JrO2d3+IFSdJK99Nk4OwwgOFp/8Xb07fnZy0EfiivT6+EgZ/wlLnCPeKfLw6jw/kEK2F8I8qOSsuHYRYw6vLziMfANMxr5e7q6HpXYBOohWRnLusQyL8cKf7Ms8vQtzkuOaCryxpAUTdKGS152fbfBTc2br/Woxw83UrsS61suW0DQlO9VBbkCL2v6ZE+bjc2Pxe0tqNf7vf4y6/iXVbOvwZft5U/je8k2BV/SxQuQXX43hj9RPQ5T5PwrZeTz94N2/cOvpL9lXK4m38/ovvgrddeH2X39IsasW9I2pVNMRxyHDVYBZQezQoWNR+y7axRpHME9+Ex8jwn464Pbb3PH7f/mGzBISC6ceO28E1tZvHqk6FOrsa3QvvP02+ruTHT6nbErXJc+uv/J1hupDKSKB2EYIeTGXqB1NE8xlE0Ya+2jeBzVRccJ2wJLJi2/GmKxI1pPLGnOp+x/6NfLcYT7D0S5dD8/3W2p4iTDbZ7WeDW4vM1fhmE4jvDRk2jDMVLyOVF1gMJ6nkpO8sU8NcYaMRerOuBlVRXGxezjfoPCTTjtXZbjT1kfvWHHUb79CMCgDnB0fjoB7BzB3vkMhlr+ltT0CShDRJliLsCdy7tKFu4Ez04AuPAAb3zyPIeB0eOOWy0Ygq2ydFCtudb1mrlMssv6/X5sVOPZriSXLsvtda3ZUn2fbR8SxhFKNrCjBjg+NtTjs+XcEop/AA)is a sandcastle demo that showcases these updates. Select "Color By Building Material" and note that line 39 is commented out.

I am having issues with the spec "setColor throws with invalid batchId". Suggestions are welcome. 

@ebogo1 
@lilleyse

cesium-concierge commented:
Thanks for the pull request @srothst1!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/3D-tile-style-exception/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

srothst1 commented:
I removed the spec "setColor throws with invalid batchId", as I do not think it is relevant. However, I would like someone to revise this decision.

lilleyse commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -1253,6 +1253,9 @@ Object.defineProperties(Cesium3DTileset.prototype, {
    * event is raised, so code in <code>tileVisible</code> can manually set a feature's
    * properties (e.g. color and show) after the style is applied. When
    * a new style is assigned any manually set properties are overwritten.
+   * Use an always "true" condition to specify the Color for all objects that are not
+   * overridden by pre-existing conditions. Otherwise, the default Color Cesium.color.White
+   * will be used.
    * </p>
```
Minor doc nitpicks

```suggestion
   * a new style is assigned any manually set properties are overwritten.
   * </p>
   * <p>
   * Use an always "true" condition to specify the Color for all objects that are not
   * overridden by pre-existing conditions. Otherwise, the default color Cesium.Color.White
   * will be used.
   * </p>
```

lilleyse commented on Source/Scene/Cesium3DTileBatchTable.js:
```diff
@@ -201,6 +201,9 @@ Cesium3DTileBatchTable.prototype.getShow = function (batchId) {
 };
 
 Cesium3DTileBatchTable.prototype.setColor = function (batchId, color) {
+  if (!defined(color)) {
+    return;
+  }
```
One problem with this approach is that if a new style is applied any features that fall into the "else" condition will stay the same color that they were given previously. I think the safer approach is to make the change directly in the `ConditionsExpression` code. 

```js
ConditionsExpression.prototype.evaluateColor = function (feature, result) {
  var conditions = this._runtimeConditions;
  if (!defined(conditions)) {
    return undefined;
  }
  var length = conditions.length;
  for (var i = 0; i < length; ++i) {
    var statement = conditions[i];
    if (statement.condition.evaluate(feature)) {
      return statement.expression.evaluateColor(feature, result);
    }
  }

  return Color.clone(Color.WHITE, result); // <------ see here
};
```
[Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=lVTbbtswDP0VIhhgB3DlpJcV7ZJiaPq4ocUabA91HxSbiYXIUiDJLrIi/z7KNzS9bBlg2DLFc3jEQ7viBiqBT2hgCgqfYIZWlAX7WcfCZJDW7zOtHBcKTTKI4DlRAA6Nocid0ZXI0Fx2wNQgd/hLG5nNm5RwGCVqN/ySqERVVM4JiRYd1duD3NriuhQyE2plQ5/dyGI2RYVsY0QhnKjQMp5lYcvRkr5MTHmBhrOl3M51WCvN0DqhuBNa9Spn3DhacXXClkYXN7gyiDY8Oj9lo/FFBKcj9vlifBzB+dnIywfQRiD1oGGpeQFy5F5uz/qdu5w5/YOiXNnwuIUCbIRL84/Sjrq8XX3PStNWGbFR37j2wMy6rcR9q5rHyc2cUu79dnPsVEtNtpCDfhEGBrOAuAY95T1XWcqtk+hbOtdaLri5Lp3Tinyf5VytEOp63vRlqVIvKxy2/v+3IO/EkmbI9v0DKMh4I7j0Oj89L2kOSoMPwaIdhMtuP3jckYgGtWuf7QF7rlSrTHiJVOChCwItibnj2cF0SlpLKf2Z+t485cJhMEwGj9FfgcFKcmuDPaxdbxeyxCAiv84OoCCZNPBUbo+F5m97iICFEel6HyqUn6Pa3n/jLbn7qrQUq9xZSXmHiijQkSVvSQ7FW4f4Hr6O1818RRLHRONMiR+5Rhkwz4UFulyOgNIi0HBjx/G4Nzz1F1DfBtFgUo/wVZf5VRQbbRyURoaMxQ6Lje+NjRdluqaBT631QJ86iV9CJ5moQGTTd/6ZkPq5oZ0lTd69+E3nuJrElP8GKnX9S7mt0Ei+9Wn5+OpbE2SMTWJ6fR/pmg/4FfMf)

Before|![3](https://user-images.githubusercontent.com/915398/123181326-18e97e80-d45b-11eb-8a80-18a940e74d03.jpg)|![4](https://user-images.githubusercontent.com/915398/123181325-18e97e80-d45b-11eb-9cad-be08a74b138a.jpg)
--|--|--
After|![3](https://user-images.githubusercontent.com/915398/123181326-18e97e80-d45b-11eb-8a80-18a940e74d03.jpg)|![2](https://user-images.githubusercontent.com/915398/123181324-18e97e80-d45b-11eb-98ea-ac80e050911a.jpg)

lilleyse replied:
Also the tests that were removed can be re-added.

lilleyse replied:
I've been thinking about how to fix this for `show` too and I don't think it's as simple to fix in `ConditionsExpression.prototype.evaluate`. Maybe the fix needs to go in `Cesium3DTileBatchTable.prototype.applyStyle`

[`show` sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=dVJda9swFP0rlzCIC66ctN1KsySMpY8bHWvYHuo8KPZNLCZL4Up26Er++6781XbrwMjSveccHR2plgS1wiMSLMDgEVboVFWKH00tSkdZs15Z46UySOkohqfUAHgk4so3srXKkWY9MSOUHn9a0vm6hURncWpOZx9Tk5qat/NKo0PP+72i3Lnyc6V0rszeRQHd2hIuQ4PiQKpUXtXohMzzqNPoRF8CM1kiSbHTj2sbNU5zdF4Z6ZU1g8uVJM8zaS7Fjmx5i3tCdNH59ZWYTG9iuJqIDzfTixiu30+CfQBLCjmDVqXRBShQBruD6lfpC+Htd65K46KLjgpwUD4r/gc773GnZswr6naZiMkQXHdg4fyjxtdX1f4ub9cMuQ/t/tg7vjA3mC05Y1JSzyAdvXvaceQV4cN422U+6/vjzYkv+dmPK+xxEMmsyVVwx7oPbQl4woI9/QSLxQLGey2dG4fX4qnCTdxjk4TRofTc4hqsC+WAP18goHYImXTYcjaDlyaKUTyaNyEse8lPqjxY8lCRjoRIPJYHzWZcsq2yXxxZ5lwgBug8eUmd56oGlS/eeOSQBf/c2VVa36vfbHc5Txj/D1Xb5g3c1UhaPgZYMV1+aYtCiHnCy7eZ3lq9lfSX8h8)

srothst1 commented:
@lilleyse I just cleaned up the `Cesium3DTileset` documentation, implemented the changes you suggested to `Cesium3DTileseBatchTable` and `ConditionsExpression`, and added the specs back to `Cesium3DTileBatchTableSpec.js`.

Fixing this bug for `show` makes sense and would prevent confusion in the future. The sandcastle demo that you provided clearly demonstrates this. Why do you think the fix needs to go in `Cesium3DTileBatchTable.prototype.applyStyle`? Any suggestions on what the fix might look like?

srothst1 commented:
The last commit seems to resolve the issue outlined in [this sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=dVJda9swFP0rlzCIC66ctN1KsySMpY8bHWvYHuo8KPZNLCZL4Up26Er++6781XbrwMjSveccHR2plgS1wiMSLMDgEVboVFWKH00tSkdZs15Z46UySOkohqfUAHgk4so3srXKkWY9MSOUHn9a0vm6hURncWpOZx9Tk5qat/NKo0PP+72i3Lnyc6V0rszeRQHd2hIuQ4PiQKpUXtXohMzzqNPoRF8CM1kiSbHTj2sbNU5zdF4Z6ZU1g8uVJM8zaS7Fjmx5i3tCdNH59ZWYTG9iuJqIDzfTixiu30+CfQBLCjmDVqXRBShQBruD6lfpC+Htd65K46KLjgpwUD4r/gc773GnZswr6naZiMkQXHdg4fyjxtdX1f4ub9cMuQ/t/tg7vjA3mC05Y1JSzyAdvXvaceQV4cN422U+6/vjzYkv+dmPK+xxEMmsyVVwx7oPbQl4woI9/QSLxQLGey2dG4fX4qnCTdxjk4TRofTc4hqsC+WAP18goHYImXTYcjaDlyaKUTyaNyEse8lPqjxY8lCRjoRIPJYHzWZcsq2yXxxZ5lwgBug8eUmd56oGlS/eeOSQBf/c2VVa36vfbHc5Txj/D1Xb5g3c1UhaPgZYMV1+aYtCiHnCy7eZ3lq9lfSX8h8) provided by @lilleyse. Modifications were made with the intention of correctly handling the "else" case for `show` - a similar fix as the one made for `color`.

lilleyse commented on Source/Scene/Cesium3DTileBatchTable.js:
```diff
@@ -236,7 +236,11 @@ Cesium3DTileBatchTable.prototype.applyStyle = function (style) {
       : DEFAULT_SHOW_VALUE;
     this.setColor(i, color);
-    this.setShow(i, show);
+    if (defined(show)) {
+      this.setShow(i, show);
+    } else {
+      this.setShow(i, false);
+    }
```
> Why do you think the fix needs to go in `Cesium3DTileBatchTable.prototype.applyStyle`? Any suggestions on what the fix might look like?

I couldn't figure out a generic way to get it working in `ConditionsExpression.prototype.evaluate` since the return type could be different types. And I think it fits cleaner in `Cesium3DTileBatchTable.prototype.applyStyle` which already has handling for default show and color. What you have is what I had in mind as well. Just a few suggestions...

I think `true` is a better default when `show` is undefined. It matches `DEFAULT_SHOW_VALUE`.

This code can be slightly more concise by using `defaultValue`. Two possible approaches:

```
    var show = defined(style.show) ? style.show.evaluate(feature) : undefined;
    show = defaultValue(show, DEFAULT_SHOW_VALUE);

    ...

    this.setShow(i, show);
```
```
    var show = defined(style.show)
      ? defaultValue(style.show.evaluate(feature), DEFAULT_SHOW_VALUE)
      : DEFAULT_SHOW_VALUE;

    ...

    this.setShow(i, show);
```

`color` could be fixed in this function as well instead of in `ConditionsExpression`. I wasn't super clear in https://github.com/CesiumGS/cesium/pull/9633#discussion_r657900751 but that idea can apply to both show and color.

lilleyse commented on CHANGES.md:
```diff
@@ -21,6 +21,7 @@
 - Fixed an error where the `positionToEyeEC` and `tangentToEyeMatrix` properties for custom materials were not set in `GlobeFS`. [#9597](https://github.com/CesiumGS/cesium/pull/9597)
 - Fixed misleading documentation in `Matrix4.inverse` and `Matrix4.inverseTransformation` that used "affine transformation" instead of "rotation and translation" specifically. [#9608](https://github.com/CesiumGS/cesium/pull/9608)
 - Fixed a regression where external images in glTF models were not being loaded with `preferImageBitmap`, which caused them to decode on the main thread and cause frame rate stuttering. [#9627](https://github.com/CesiumGS/cesium/pull/9627)
+- Fixed misleading "else" case condition for `Cesium3DTileStyle`. A default color value is used if no color conditions are given. [#9633](https://github.com/CesiumGS/cesium/pull/9633)
```
`show` should be mentioned now as well

lilleyse commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -1254,6 +1254,11 @@ Object.defineProperties(Cesium3DTileset.prototype, {
    * properties (e.g. color and show) after the style is applied. When
    * a new style is assigned any manually set properties are overwritten.
    * </p>
+   * <p>
+   * Use an always "true" condition to specify the Color for all objects that are not
+   * overridden by pre-existing conditions. Otherwise, the default color Cesium.Color.White
+   * will be used.
+   * </p>
```
Mention `show` here as well

lilleyse reviewed (commented):
@srothst1 thanks for the update. Just a few final comments from me.

srothst1 commented:
@lilleyse I believe I got to all of your suggestions:
1. made the default value for `show` `true`, matching `DEFAULT_SHOW_VALUE`
2. removed `Color` references from `ConditionsExpression.js`
3. made code in `Cesium3DTileBatchTable.prototype.applyStyle.js` more concise
4. updated `CHANGES.md` with information about `show`
5. updated documentation in `Cesium3DTileset.js` to include information about `show`

Let me know what you think!

lilleyse commented:
The code changes look great! Nothing else to add there.

I should have mentioned this earlier - could you add two tests to `Cesium3DTileset`?

You can model the tests after `applies conditional color style to a tileset` though I don't think you need to check the pixel colors like those tests do. Instead it might be something like:

* Load the `withBatchTableUrl` tileset and add a style that sets the color of buildings with `${id} > 0` to red (as a condition expression with just one condition)
* Check that feature 0's color is white and feature 1's color is red. You can access the feature with `tileset.root.content.getFeature(0)`

The test for the `show` property would look similar.

srothst1 commented:
@lilleyse I just added the two tests to `Cesium3DTileset`.

lilleyse commented on Specs/Scene/Cesium3DTilesetSpec.js:
```diff
@@ -2901,6 +2901,53 @@ describe(
+              conditions: [["${id} > 0", 'color("black")']],
+            },
+          });
+          expect(
+            tileset.root.content.getFeature(0).hasProperty('color("white")')
+          );
+          expect(
+            tileset.root.content.getFeature(1).hasProperty('color("black")')
+          );
```
These two checks are actually no-ops. `expect(.......).toBe(true)` needs to be added. After making that change you'll see that the tests fail.

`hasProperty` is meant for querying properties that are stored in a content's [batch table](https://github.com/CesiumGS/3d-tiles/tree/master/specification/TileFormats/BatchTable). You can see the full list of properties when you click a feature in CesiumJS, e.g. click a building in [3D Tiles Adjust Height](https://sandcastle.cesium.com/?src=3D%20Tiles%20Adjust%20Height.html&label=3D%20Tiles) and it will show `id`, `Longitude`, `Latitude`, and `Height`.

`getFeature(0).hasProperty("id")` would return true but `getFeature(0).hasProperty('color("white")')` would not, unless for some reason the 3D Tiles had a property called `color("white")`.

To get a feature's color or show, which are actually CesiumJS specific constructs, you can use the `color` and `show` getters.

lilleyse replied:
And instead of `.toBe(true)` something like this is better: 

`expect(tileset.root.content.getFeature(0).color).toEqual(Color.WHITE);`

srothst1 replied:
@lilleyse I had originally written the spec using code similar to your recent suggestions: 

spec: `"handle else case when applying conditional color style to a tileset"`
`expect(tileset.root.content.getFeature(0).color).toEqual(Color.WHITE);`
`expect(tileset.root.content.getFeature(1).color).toEqual(Color.BLACK);`

and 

spec: `"handle else case when applying conditional show to a tileset"`
`expect(tileset.root.content.getFeature(0).show).toBe(true);`
`expect(tileset.root.content.getFeature(1).show).toBe(false);`

However, I noticed that both specs failed even though my sandcastle demos seemed to have the desired functionality. It seems like the `if` case conditions are not being applied. Any suggestions?

lilleyse replied:
Add `scene.renderForSpecs();` after setting the style and the test should pass. The style will be applied during the update.

lilleyse replied:
The styles in `handle else case when applying conditional show to a tileset` need to be fixed. `show` does not have a constructor in the styling language like `color` does.

See https://github.com/CesiumGS/3d-tiles/tree/master/specification/Styling#styling-features. Just `"true"` and `"false"` alone should work.

srothst1 replied:
@lilleyse I updated the specs:
`"handle else case when applying conditional color style to a tileset"`
and 
`"handle else case when applying conditional show to a tileset"`.

It seems like adding `scene.renderForSpecs();` after setting the style was the missing piece. One other thing, in `"handle else case when applying conditional show to a tileset"` we see that the else case defaults to `false` - is this what we envisioned?

lilleyse replied:
Ah that's because the new `show` tests are not using condition expressions

srothst1 replied:
@lilleyse We should be good to go now. I apologize for this last step taking so long. Thank you very much for your help!

lilleyse replied:
No problem, all part of the learning experience 🙂
